### PR TITLE
[WIP]: Fix array growth thresholding

### LINKED ISSUE #8269: Grow arrays by a factor less than 2?
I recently discovered this document in one of FB’s open source libraries for C++: https://github.com/facebook/folly/blob/master/folly/docs/FBVector.md

It makes an interesting argument for growing arrays by a factor smaller than 2 when resizing them so that you can reuse memory more easily. Not sure how much you also need to use a custom malloc routine to get the benefits being claimed, but I thought I'd raise the idea in case we can get any performance enhancements from these ideas.

### PR DESCRIPTION
**Pull request by NHDaly**
To address problem (1) in https://github.com/JuliaLang/julia/issues/28588#issuecomment-492699026: Fixing the quadratic total insertion time complexity.

In this PR, currently I just always grow at a rate of 1.5x (after the number of elements is greater than 10). Happy to explore other options!

Closes #8269.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
It seems to me there should still be some kind of limit; 1% of RAM is just really stingy.

**Comment by NHDaly** [comment]
In case you didn't see it, I've summarized my latest thoughts in this comment:
https://github.com/JuliaLang/julia/issues/28588#issuecomment-492699026

My main concern is that there should never be a constant-size growth increment; it should always be a scaling growth factor. As long as we do that, I'm happy.

I think we could consider lowering that growth factor based on the array size (to address your suggestion), but doing will still (I think) have implications on the time complexity. And given the benefits of virtual memory, it doesn't seem to actually be so terrible to let arrays grow larger than physical RAM anyway. Interested to hear your thoughts! :)

**Comment by JeffBezanson** [comment]
Ok, that's a good argument. I'm ok with just doing 1.5x growth for now. I'd also really like to use realloc again. Large allocations are highly likely to be well aligned, so the try-and-check approach seems safe enough to me.

**Comment by StefanKarpinski** [comment]
Isn't the golden ratio supposed to be the optimal growth factor?

**Comment by JeffBezanson** [comment]
Yes, the golden ratio is optimal, and I suppose we don't mind using floating point, so we could use it. I wonder why it's not often used in practice though. Maybe you want some wiggle room in case the new size doesn't exactly fit for implementation reasons, or in case part of the old space has been allocated to something else?

**Comment by KristofferC** [comment]
Ref https://github.com/JuliaLang/julia/pull/16305

**Comment by JeffBezanson** [comment]
Ah yes. That discussion also reminds me that we should grow faster (e.g. 2x) up to some not-too-big, not-too-small threshold like 1000. Under that size wasting memory doesn't matter as much, but doing less reallocation will probably bring measurable speedups.

**Comment by vtjnash** [comment]
Yes, fwiw, we discussed during JuliaCon using some other growth function like `n + k sqrt n` or `n + k log n`, because they can be made to exhibit the behavior you describe but with smoother cut-off points. It's then straight forward to pick the cross-over point such that it strictly superior to the current exponential growth strategy for all finite array sizes.

**Comment by oscardssmith** [comment]
@JeffBezanson The main reason the golden ratio is typically not used is that arrays are not that frequently grown, and if anything gets allocated in between it would go badly. IIRC, python uses 1.125 because they looked at a bunch of real world code and found that repeated pushes aren't that frequent.

**Comment by JeffBezanson** [comment]
I agree that it's probably rare to do lots of unpredictable pushes to an array over its lifetime, but the case I worry about is where `push!` is used to initially populate an array. In that case growing too slowly to the final size can hurt performance quite a bit. 1.125 seems a bit too small for that case.

**Comment by StefanKarpinski** [comment]
How about using powers of two until the array size is an entire OS page? That should help avoid fragmentation since powers of two are easier to pack. Once the array size is ≥ a page, one can use any growth fact at all since it will always be in terms of whole pages.

**Comment by NHDaly** [comment]
> @JeffBezanson The main reason the golden ratio is typically not used is that arrays are not that frequently grown, and if anything gets allocated in between it would go badly. IIRC, python uses 1.125 because they looked at a bunch of real world code and found that repeated pushes aren't that frequent.

Makes sense. Thanks for the explanation! :)

> Yes, fwiw, we discussed during JuliaCon using some other growth function like `n + k sqrt n` or `n + k log n`, because they can be made to exhibit the behavior you describe but with smoother cut-off points. It's then straight forward to pick the cross-over point such that it strictly superior to the current exponential growth strategy for all finite array sizes.

@vtjnash yeah, i was really excited about that at the time too. But after thinking about it more, I now think it's a bad idea, per the reasons i wrote in https://github.com/JuliaLang/julia/issues/28588#issuecomment-492699026, section "Changing growth factor based on physical RAM": Basically i think that if you _shrink_ (even slightly) the size you grow by after each growth, you will end up with an ammortized insertion time bigger than `O(1)`, I think probably `O(n)`. I think you need to keep the growth rate constant in order to stay ammortized constant insert time.

> How about using powers of two until the array size is an entire OS page? That should help avoid fragmentation since powers of two are easier to pack and it makes the growth for small dicts higher, addressing what Jeff is concerned about. Once the array size is ≥ a page, one can use any growth fact at all since it will always be in terms of whole pages.

This seems like a reasonable idea to me! I'll think about it more.

-----------

Also, lemme add @tveldhui for some more input. He's thought about this a bit as well.

**Comment by tveldhui** [comment]
Julia is designed to be high performance.  To my mind, doubling the array size is most consistent with high performance, it reduces copies per element in situations where realloc doesn't just do MREMAP.

The physical memory limit seems a red herring.  There aren't many people who are going to accidentally coincidentally create one single array that approaches the size of physical memory.  Most people working on big interesting problems have many data structures of varying sizes.

But consider the case where someone does have an interesting problem that requires one huge vector. If you picture a log-scale graph of problem sizes, the ones right around physical memory size are a razor-thin transition.  If your problem fits in memory, you want to minimize copying for performance reasons. As you approach physical memory size you've got swap and overcommit to fall back on.  If you're tackling problems bigger than main memory and determined to use a single Array/Vector, you're going to be relying on NVMe swap or somesuch.

From what I can tell, the main motivation for picking a smaller growth size than 2x, or having one that trails off, is to allow VERY LARGE arrays get _slightly closer_ to the physical RAM limit without OOMing on machines that don't have any swap available. This seems like a very rare use-case to me, and to me, it doesn't make sense to penalize 99.9% of use cases by choosing a lower-performance resizing heuristic to help a tiny fraction of users tackle marginally bigger problems before they OOM.

**Comment by NHDaly** [comment]
Ah, yeah, that's a good point. My main motivation for suggesting 1.5x instead of 2x was to avoid this packing problem leaving holes in memory. But I guess if we tackle this at the same time as fixing `realloc` to actually do realloc, then the packing problem might not even be a real problem anymore, in which case I'm also down for just constant 2x growth.


I'll try to add the realloc fix to this PR, and then maybe get some graphs of memory use and CPU use for different growth factors?

**Comment by oscardssmith** [comment]
@tveldhui the counter argument would be that frequent array growth is rare in performance critical work. (And if it matters, you can always manually set capacity). As a result the place memory efficiency matters most is large numbers of small arrays, which to me implies a fairly small growth factor consistently.

**Comment by NHDaly** [comment]
Oscar, for many small vectors I would feel even more strongly about wanting
a larger growth factor.


There's obviously a CPU performance vs memory tradeoff: a smaller growth
rate would grow more often (more CPU expensive) in exchange for wasting
less memory from overallocating.

But from my perspective, the memory gains aren't that significant from a
smaller growth rate. In the 2x growth scenario, if you assume the absolute
worst case, every vector would be exacty twice as big as it should be.

So in the absolute best case from a memory perspective, you only have a
constant factor improvement to make by using a smaller rate.

But you pay the cpu cost _every time_ you grow the array, so the potential
cost there is large. And smaller arrays necessarily grow more often. So
with many small arrays a lot of your CPU would be spent in the allocator.

So for many small arrays, I would favor a larger growth rate even more
strongly, since I wouldn't want to be regrowing them all the time every few
inserts.

Does that seem reasonable?

**Comment by timholy** [comment]
@oscardssmith 

> IIRC, python uses 1.125 because they looked at a bunch of real world code and found that repeated pushes aren't that frequent.

I doubt that's directly translatable. In a slow language like Python you don't want to create arrays element-by-element because the interpreter cost kills you---you do everything you can to create arrays all at once. Julia doesn't have that overhead, so there is less disincentive to use `push!`. FWIW, there are currently 619 uses of `push!` in `base/` alone; I didn't take the time to see how many of them occur inside a loop, but I bet it's not small.

**Comment by tveldhui** [comment]
> @tveldhui the counter argument would be that frequent array growth is rare in performance critical work. (And if it matters, you can always manually set capacity). As a result the place memory efficiency matters most is large numbers of small arrays, which to me implies a fairly small growth factor consistently.

We're implementing an ultra high performance database engine/machine learning system in Julia.  I work on database queries. (And for the record I am in love with Julia for this, we are outperforming the very best database engines single-core.) As you evaluate a query you're often doing vector push! to accumulate the results.  With meticulous sampling and hairy algorithms you can approximate the expected result size, but that doesn't help you set an appropriate size for the vector of results - if your estimate is slightly under you still have to resize.  Or you can go for an upper bound on the result size, which is computationally intensive and can be off by orders of magnitude (e.g. you reserve 100 times more space than needed).  Database theory is a sticky wicket.

So I disagree very much that frequent array growth is rare and that you can predetermine capacity.  Most of our cpu time is spent in queries where we cannot predict in advance how many results will be produced, so a fast push! is critical.

File: src/array.c [hunk 1]
```diff
@@ -695,16 +693,9 @@ static void NOINLINE array_try_unshare(jl_array_t *a)
...
 {
-    // Limit overallocation to jl_arr_xtralloc_limit
-    size_t es = a->elsize;
-    size_t xtra_elems_mem = (newlen - a->offset - alen - inc) * es;
-    if (xtra_elems_mem > jl_arr_xtralloc_limit) {
-        // prune down
-        return alen + inc + a->offset + (jl_arr_xtralloc_limit / es);
-    }
-    return newlen;
+    return curlen <= 10 ? curlen * 2 : curlen * 1.5;
```
Comment (stevengj) [inline src/array.c:698]: Not that it matters much, but in #16305 I used `((curlen*3)>>1)` to avoid floating-point conversion here.

Comment (musm) [inline src/array.c:698]: Should the threshold  be increased to 1000 as per Jeff's comment?

**Comment by musm** [comment]
@NHDaly any interest in picking this up again? Almost all the arugments are in favor of growing the arrays by a factor less than 2, and based on this and the previous PR a 1.5x growth factor was deemed a resonably compromise.

**Comment by eschnett** [comment]
I hate to be that guy, but I notice that none of the arguments presented here are supported by either memory usage or run time benchmarks...

Benchmarks could include building Julia, running the test cases, or some sparse matrix algebra that might use `push!` to populate sparse matrices.

**Comment by KristofferC** [comment]
I think this is closed via https://github.com/JuliaLang/julia/pull/40453.

**Pull request closed without merging**